server/apps/immich/src/utils/disk-storage.ts

```diff
@@ -0,0 +1,101 @@
+import { Request } from 'express';
+import { ParamsDictionary } from 'express-serve-static-core';
+import { DiskStorageOptions, StorageEngine } from 'multer';
+import * as crypto from 'node:crypto';
+import * as fs from 'node:fs';
+import * as os from 'node:os';
+import * as path from 'node:path';
+import { PassThrough } from 'node:stream';
+import type { ParsedQs } from 'qs';
+
+type GetFileNameFn = (req: Request, file: Express.Multer.File, cb: (error: Error | null, filename: string) => void) => void;
+type GetDestFn = (req: Request, file: Express.Multer.File, cb: (error: Error | null, destination: string) => void) => void;
+
+export interface ExtendedDiskStorageOptions extends DiskStorageOptions {
+}
+
+function getFilename(
+  req: Request,
+  file: Express.Multer.File,
+  cb: (error: Error | null, filename: string) => void
+) {
+  crypto.randomBytes(16, function (err, raw) {
+    cb(err, err ? '' : raw.toString('hex'));
+  });
+}
+
+function getDestination(
+  req: Request,
+  file: Express.Multer.File,
+  cb: (error: Error | null, destination: string) => void
+) {
+  cb(null, os.tmpdir())
+}
+
+class ExtendedDiskStorage implements StorageEngine {
+  private _getFilenameFn: GetFileNameFn;
+  private _getDestinationFn: GetDestFn;
+
+  constructor(opts: ExtendedDiskStorageOptions) {
+    this._getFilenameFn = (opts.filename || getFilename);
+
+    if (typeof opts.destination === 'string') {
+      fs.mkdirSync(opts.destination, { recursive: true });
+      this._getDestinationFn = ($0, $1, cb) => {
+        cb(null, opts.destination as string);
+      };
+    } else {
+      this._getDestinationFn = (opts.destination || getDestination);
+    }
+  }
+
+  public _handleFile(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, file: Express.Multer.File, cb: (error?: any, info?: Partial<Express.Multer.File> | undefined) => void): void {
+    const that = this;
+
+    that._getDestinationFn(req, file, (err, destination) => {
+      if (err) return cb(err);
+
+      that._getFilenameFn(req, file, (err, filename) => {
+        if (err) return cb(err);
+
+        const finalPath = path.join(destination, filename);
+        const outStream = fs.createWriteStream(finalPath);
+        const passThrough = new PassThrough();
+        const sha1Hash = crypto.createHash('sha1').setEncoding('hex');
+        const deferred$ = new Promise<string>((resolve, reject) => {
```

Comment:
Why do we use a variable with `$` symbol here?

Reply:
just convention for observer object (eg: deffer, Observer from RxJS)

Reply:
But we don't use the notion of the observable object in our code though

Reply:
I think we should a coding guideline to avoid this issue. I will remove `$` for unified code style.

Reply:
any good guidelines for Nestjs that you'd recommend?

Reply:
nowadays, I follow Airbnb coding style and combine with Angular's one. IMO it depend on our team member or just from you 😆 I will follow all you.

Reply:
We have a coding style 😃 , we use prettier and standard `@typescript-eslint`. You can configure your IDE with ESLint / Prettier to get instant feedback, but can also run `npm run check:all` which does type checking and lint (we should get this command added to the CI).
`that` is not needed when using arrow functions since they don't create new scopes. `that` aliasing would get picked up by the linter as an error: `Unexpected aliasing of 'this' to local variable`.

The keyword `public` is not needed in method `_handleFile`, the linter would also warn on that. Also, since is public, by convention it shouldn't have an underscore.

I`_handleFile` method is exceeding 120 characters (max configured in prettier), if you get prettier configured in your IDE so it formats on save, you don't have to worry about these ones, they get formatted automatically when you save the file.

I think we should use Promises instead of callbacks (and async/await). We can use callbacks internally but return a Promise 😄

Reply:
Hi @jbaez, thanks for your comment. FYI, my approach is `implements StorageEngine`  so some functions like `_handleFile` we must override from base. I also refer to use async/await the same as you. About lint, I don't catch any error with `$` notation. For me, this just about guideline/rule that out of linter work.

Reply:
@panoti yeah, I just had a quick look and missed that is implementing `StorageEngine`, so yeah if StorageEngine defines a public method with `_` there is nothing we can do 😃.
We don't have any linter rule about `$`, but yeah as you say this one would be more like a guideline instead of an enforced lint rule.